src/Servers/Kestrel/Transport.MsQuic/src/Internal/MsQuicStream.cs

```diff
@@ -0,0 +1,241 @@
+            }
+        }
+
+        public PipeWriter Input => Application.Output;
+
+        public PipeReader Output => Application.Input;
+
+        public uint HandleStreamEvent(
+            QuicStream stream,
+            ref MsQuicNativeMethods.StreamEvent evt)
+        {
+            var status = MsQuicConstants.Success;
+
+            switch (evt.Type)
+            {
+                case QUIC_STREAM_EVENT.START_COMPLETE:
+                    status = HandleStartComplete(stream);
+                    break;
+                case QUIC_STREAM_EVENT.RECV:
+                    {
+                        status = HandleEventRecv(
+                            stream,
+                            ref evt);
+                    }
+                    break;
+                case QUIC_STREAM_EVENT.SEND_COMPLETE:
+                    {
+                        status = HandleEventSendComplete(ref evt);
+                    }
+                    break;
+                case QUIC_STREAM_EVENT.PEER_SEND_CLOSE:
+                    {
+                        status = HandleEventPeerSendClose();
+                    }
+                    break;
+                // TODO figure out difference between SEND_ABORT and RECEIVE_ABORT
+                case QUIC_STREAM_EVENT.PEER_SEND_ABORT:
+                    {
+                        _streamClosedTokenSource.Cancel();
+                        status = HandleEventPeerSendAbort();
+                    }
+                    break;
+                case QUIC_STREAM_EVENT.PEER_RECV_ABORT:
+                    {
+                        _streamClosedTokenSource.Cancel();
+                        status = HandleEventPeerRecvAbort();
+                    }
+                    break;
+                case QUIC_STREAM_EVENT.SEND_SHUTDOWN_COMPLETE:
+                    {
+                        status = HandleEventSendShutdownComplete(ref evt);
+                    }
+                    break;
+                case QUIC_STREAM_EVENT.SHUTDOWN_COMPLETE:
+                    {
+                        _stream.Close();
+                        return MsQuicConstants.Success;
+                    }
+
+                default:
+                    break;
+            }
+            return status;
+        }
+
+        private uint HandleEventPeerRecvAbort()
+        {
+            return MsQuicConstants.Success;
+        }
+
+        private uint HandleEventPeerSendAbort()
+        {
+            return MsQuicConstants.Success;
+        }
+
+        private uint HandleStartComplete(QuicStream stream)
+        {
+            return MsQuicConstants.Success;
+        }
+
+        private uint HandleEventSendShutdownComplete(ref MsQuicNativeMethods.StreamEvent evt)
+        {
+            return MsQuicConstants.Success;
+        }
+
+        private uint HandleEventPeerSendClose()
+        {
+            // TODO complete async
+            // Close as fast as possible here.
+            Input.Complete();
```

Comment:
Does this happen after *all* reads occur? That is will HandleEventRecv ever be called after this?

Reply:
I think that's an invariant but can confirm.

Reply:
It's guaranteed to come after all receives.